source/fuzz/transformation_replace_add_sub_mul_with_carrying_extended.cpp

```diff
@@ -0,0 +1,194 @@
+                                                       uint32_t result_id) {
+  message_.set_struct_fresh_id(struct_fresh_id);
+  message_.set_struct_type_id(struct_type_id);
+  message_.set_result_id(result_id);
+}
+
+bool TransformationReplaceAddSubMulWithCarryingExtended::IsApplicable(
+    opt::IRContext* ir_context, const TransformationContext&) const {
+  // |message_.struct_fresh_id| must be fresh.
+  if (!fuzzerutil::IsFreshId(ir_context, message_.struct_fresh_id())) {
+    return false;
+  }
+
+  // |message_.result_id| must refer to a suitable OpIAdd, OpISub or OpIMul
+  // instruction.
+  auto instruction =
+      ir_context->get_def_use_mgr()->GetDef(message_.result_id());
+  auto instruction_opcode = instruction->opcode();
+
+  switch (instruction_opcode) {
+    case SpvOpIAdd:
+    case SpvOpISub:
+      if (!TransformationReplaceAddSubMulWithCarryingExtended::
+              IsAddSubInstructionSuitable(ir_context, instruction))
+        return false;
+      break;
+    case SpvOpIMul:
+      if (!TransformationReplaceAddSubMulWithCarryingExtended::
+              IsMulInstructionSuitable(ir_context, instruction))
+        return false;
+      break;
+    default:
+      return false;
+  }
+  return true;
+}
+
+void TransformationReplaceAddSubMulWithCarryingExtended::Apply(
+    opt::IRContext* ir_context, TransformationContext* /*unused*/) const {
+  // |message_.struct_fresh_id| must be fresh.
+  assert(fuzzerutil::IsFreshId(ir_context, message_.struct_fresh_id()) &&
+         "|message_.struct_fresh_id| must be fresh");
+
+  auto original_instruction =
+      ir_context->get_def_use_mgr()->GetDef(message_.result_id());
+  auto original_instruction_opcode = original_instruction->opcode();
+  auto operand_type_id = original_instruction->type_id();
+  uint32_t operand_signedness =
+      ir_context->get_def_use_mgr()
+          ->GetDef(original_instruction->type_id())
+          ->GetSingleWordOperand(kOpTypeIndexSignedness);
+
+  fuzzerutil::UpdateModuleIdBound(ir_context, message_.struct_fresh_id());
+
+  // Determine the opcode of the new instruction that computes the result into a
+  // struct.
+  SpvOp new_instruction_opcode;
+
+  switch (original_instruction_opcode) {
+    case SpvOpIAdd:
+      new_instruction_opcode = SpvOpIAddCarry;
+      break;
+    case SpvOpISub:
+      new_instruction_opcode = SpvOpISubBorrow;
+      break;
+    case SpvOpIMul:
+      if (operand_signedness == 0) {
+        new_instruction_opcode = SpvOpUMulExtended;
+      } else {
+        new_instruction_opcode = SpvOpSMulExtended;
+      }
+      break;
+    default:
+      assert(false);
+      new_instruction_opcode = SpvOpMax;
+      // No uninitialized variables left (to keep compilers happy).
+      break;
+  }
+
+  // Insert the OpCompositeExtract. This instruction takes the first component
+  // of the struct which represents low-order bits of the operation. This the
+  // original result.
+  auto instruction_composite_extract =
```

Comment:
Nit: consider inserting the new Add/Sub/Mul instruction first, before the original instruction. Then insert the extract, also before the original instruction.